### ISSUE TITLE
Interaction of private Git VCS, cache, and credentials

### ISSUE THREAD

**Issue by kelleyjd** [composer/composer#8293]
We're seeing one or two issues from using a private Git repository with Composer. This is using Composer 1.9.0 with PHP 7.2.21 on RHEL 7.7.

We have a private Composer package, stored in a [private Git repository](https://getcomposer.org/doc/05-repositories.md#using-private-repositories). Our package.json looks something like this:

```
{
    "require": {
        "myorg/private-package": "^1.2"
    },
    "repositories": [
        {
            "type": "vcs",
            "url": "https://bitbucket.example.com/scm/myorg/private-package.git"
        }
    ]
}
```

*Potential issue 1*: Running `composer i` creates a cached copy of the Git repository under `~/.cache/composer/vcs` or `~/.composer/cache/vcs`, as expected. However, it stores the credentials used within the `origin` field of that cached Git repository's `config` file. We pass in our credentials through COMPOSER_AUTH or enter them from the command-line prompt to try to avoid storing them locally on the server, so finding out that they're cached in cleartext anyway is surprising and possibly undesirable.

Can the code be updated to not store credentials here? (I haven't looked at the Git driver to see how practical this is.) Alternatively, can the docs be updated to make this potential issue explicit?

*Potential issue 2*: We periodically get errors when we try to install packages:

```
Package operations: 84 installs, 0 updates, 0 removals
  - Installing myorg/private-package (1.2.2): Cloning abcdef from cache
    abcdef is gone (history was rewritten?)

  [RuntimeException]
  Failed to execute git checkout 'abcdef' -- && git reset --hard 'abcdef' --

  fatal: reference is not a tree: abcdef
```

From looking into this, I *think* that the problem is that, if Composer is unable to update its private repository cache (e.g., because the credentials that it saved within the `origin` field, as described under issue 1, are no longer valid), it silently falls back to using the out-of-date repository.

Can Composer be updated to re-prompt for credentials, even if a cached copy is available? Alternatively, can it show a warning if it can't update a cached repository?

Thank you.

**Comment by sseypt** [comment]
We have the same problem in a ci / gitlab runner environment using composer 1.8.6

tldr: The problem is that `git remote set-url origin` changes the remote repo url in the vendor directory to the version without authentication. Although authentication is previously used in the local cache repo.

This is done with this [$command](https://github.com/composer/composer/blob/19b5f66a0e233eb944f134df34091fe1c5dfcc11/src/Composer/Downloader/GitDownloader.php#L60).

The problem is this [$condition](https://github.com/composer/composer/blob/19b5f66a0e233eb944f134df34091fe1c5dfcc11/src/Composer/Util/Git.php#L97).

`$this->process->execute` would normaly fail with an origin without authentication.
As it does in [syncMirror()](https://github.com/composer/composer/blob/19b5f66a0e233eb944f134df34091fe1c5dfcc11/src/Composer/Util/Git.php#L233).

But the [$command](https://github.com/composer/composer/blob/19b5f66a0e233eb944f134df34091fe1c5dfcc11/src/Composer/Downloader/GitDownloader.php#L60) doesn't fail, so the origin remains without authentication.

Hope that helps.

# setup / ci

```
cd /builds/myorg
composer config --auth -- http-basic.example.com gitlab-ci-token ${CI_JOB_TOKEN}
composer install --no-interaction -vvv
```

# logs

## first run / cache empty

```
- Installing myorg/private-package (1.0.0): 
    Cloning to cache at '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/'
Executing command (CWD): git clone --mirror 'https://example.com/private-package-dist.git' '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/'
Executing command (CWD): git clone --mirror 'https://gitlab-ci-token:***@example.com/private-package-dist.git' '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/'
Cloning e59123cd45dd1bdfa7da897d523f37c5203415c7 from cache
Executing command (CWD): git clone --no-checkout '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/' '/builds/myorg/vendor/myorg/private-package' --dissociate --reference '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/' && cd '/builds/myorg/vendor/myorg/private-package' && git remote set-url origin 'https://example.com/private-package-dist.git' && git remote add composer 'https://example.com/private-package-dist.git'
Executing command (/builds/myorg/vendor/myorg/private-package): git branch -r
Executing command (/builds/myorg/vendor/myorg/private-package): git checkout '1.0.0' --
Executing command (/builds/myorg/vendor/myorg/private-package): git reset --hard 'e59123cd45dd1bdfa7da897d523f37c5203415c7' --
    REASON: Required by the root package: Install command rule (install myorg/private-package 1.0.0)
```

## same version / from cache

```
- Installing myorg/private-package (1.0.0): 
    Cloning to cache at '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/'
Executing command (/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/): git rev-parse --git-dir
Executing command (/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/): git rev-parse --quiet --verify 'e59123cd45dd1bdfa7da897d523f37c5203415c7^{commit}'
Cloning e59123cd45dd1bdfa7da897d523f37c5203415c7 from cache
Executing command (CWD): git clone --no-checkout '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/' '/builds/myorg/vendor/myorg/private-package' --dissociate --reference '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/' && cd '/builds/myorg/vendor/myorg/private-package' && git remote set-url origin 'https://example.com/private-package-dist.git' && git remote add composer 'https://example.com/private-package-dist.git'
Executing command (/builds/myorg/vendor/myorg/private-package): git branch -r
Executing command (/builds/myorg/vendor/myorg/private-package): git checkout '1.0.0' --
Executing command (/builds/myorg/vendor/myorg/private-package): git reset --hard 'e59123cd45dd1bdfa7da897d523f37c5203415c7' --
    REASON: Required by the root package: Install command rule (install myorg/private-package 1.0.0)
```

## new version / from cache

```    
- Installing myorg/private-package (2.0.0): 
    Cloning to cache at '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/'
Executing command (/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/): git rev-parse --git-dir
Executing command (/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/): git rev-parse --quiet --verify 'cf9f1f00dd277910d5df9cfe8adf5aeb8178a9d0^{commit}'
Executing command (/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/): git rev-parse --git-dir
Executing command (/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/): git remote -v
Executing command (/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/): git remote set-url origin 'https://example.com/private-package-dist.git' && git remote update --prune origin
Executing command (/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/): git remote set-url origin 'https://gitlab-ci-token:***@example.com/private-package-dist.git' && git remote update --prune origin
Executing command (CWD): git --version
Cloning cf9f1f00dd277910d5df9cfe8adf5aeb8178a9d0 from cache
Executing command (CWD): git clone --no-checkout '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/' '/builds/myorg/vendor/myorg/private-package' --dissociate --reference '/cache/.composer/cache/vcs/https---example.com-private-package-dist.git/' && cd '/builds/myorg/vendor/myorg/private-package' && git remote set-url origin 'https://example.com/private-package-dist.git' && git remote add composer 'https://example.com/private-package-dist.git'
Executing command (/builds/myorg/vendor/myorg/private-package): git branch -r
Executing command (/builds/myorg/vendor/myorg/private-package): git checkout '2.0.0' --
Executing command (/builds/myorg/vendor/myorg/private-package): git checkout -B '2.0.0' 'composer/2.0.0' --
Executing command (/builds/myorg/vendor/myorg/private-package): git checkout 'cf9f1f00dd277910d5df9cfe8adf5aeb8178a9d0' -- && git reset --hard 'cf9f1f00dd277910d5df9cfe8adf5aeb8178a9d0' --
    cf9f1f00dd277910d5df9cfe8adf5aeb8178a9d0 is gone (history was rewritten?)
Failed: [RuntimeException] Failed to execute git checkout 'cf9f1f00dd277910d5df9cfe8adf5aeb8178a9d0' -- && git reset --hard 'cf9f1f00dd277910d5df9cfe8adf5aeb8178a9d0' --

fatal: reference is not a tree: cf9f1f00dd277910d5df9cfe8adf5aeb8178a9d0
```

Cheers, Steve
